File: StriverA2Z/Java/Arrays/Medium/SetMatrixZero.java
public class SetMatrixZero {
    public static void main(String[] args) {
        int[][] arr = {{1, 2, 3, 1}, {4, 0, 2, 6}, {3, 4, 0, 2}, {0, 3, 1, 5}};
        setMatrixZero(arr);
    }

    static void printMatrix(int[][] arr) {
        for (int i = 0; i < arr.length; i++) {
            for (int j = 0; j < arr[0].length; j++) {
                System.out.print(arr[i][j] + " ");
            }
            System.out.println();
        }
        System.out.println();
    }

    static void setMatrixZero(int[][] arr) {
        printMatrix(arr);
        int col0 = 1;
        int n = arr.length;
        int m = arr[0].length;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (arr[i][j] == 0) {
                    arr[i][0] = 0;
                    if (j != 0) arr[0][j] = 0;
                    else col0 = 0;
                }
            }
        }

        for (int i = 1; i < n; i++) {
            for (int j = 1; j < m; j++) {
                if (arr[i][j] != 0) {
                    if (arr[i][0] == 0 || arr[0][j] == 0) {
                        arr[i][j] = 0;
                    }
                }
            }
        }

        if (arr[0][0] == 0) {
            for (int j = 0; j < m; j++) {
                arr[0][j] = 0;
            }
        }

        if (col0 == 0) {
            for (int i = 0; i < m; i++) {
                arr[i][0] = 0;
            }
        }

        printMatrix(arr);
    }
}

File: StriverA2Z/C++/3_ArraysQuestion/3.2_Medium/set_matrix_zero.cpp
#include<bits/stdc++.h>

using namespace std;


void printMatrix(vector<vector<int>> vec) {
    for (int i = 0; i < vec.size(); i++) {
        for (int j = 0; j < vec[0].size(); j++) {
            cout << vec[i][j] << " ";
        }
        cout << endl;
    }
    cout << endl;
}


void set_zero(vector<vector<int>> &matrix) {
    printMatrix(matrix);

    int col0 = 1;
    int m = matrix.size();
    int n = matrix[0].size();

    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            if (matrix[i][j] == 0) {
                matrix[i][0] = 0;
                if (j != 0) matrix[0][j] = 0;
                else col0 = 0;
            }
        }
    }
    for (int i = 1; i < m; i++) {
        for (int j = 1; j < n; j++) {
            if (matrix[i][j] != 0) {
                if (matrix[i][0] == 0 || matrix[0][j] == 0) {
                    matrix[i][j] = 0;
                }
            }
        }
    }

    if (matrix[0][0] == 0) {
        for (int j = 0; j < n; j++) {
            matrix[0][j] = 0;
        }
    }

    if (col0 == 0) {
        for (int i = 0; i < m; i++) {
            matrix[i][0] = 0;
        }
    }

    printMatrix(matrix);
}

int main() {
    vector<vector<int>> vec = {{1, 2, 3, 1},
                               {4, 0, 2, 6},
                               {3, 4, 0, 2},
                               {0, 3, 1, 5}};
    set_zero(vec);
    return 0;
}